### PR TITLE
Nuclear bombs use TGUI and world.time

### PR DESCRIPTION
**Pull request by coiax**
:cl: coiax
add: Nuclear bombs now use a more accurate timer system, the same as gang dominators and shuttles.
add: Nuclear bombs now have a TGUI interface.
/:cl:

image
- Removes the fact that self destruct nukes are for some reason really high on the layer list?
- Makes some use of my wonderful admin message macros
- ~~Station nuke and syndicate nuke use different styles.~~
- Both use NT styles.

### CODE REVIEW

**Comment by KorPhaeron** [comment]
tgui was a mistake

**Comment by coiax** [comment]
?

File: code/game/gamemodes/nuclear/nuclearbomb.dm [hunk 1]
```diff
@@ -209,102 +225,129 @@ var/bomb_set
...
-							code = "ERROR"
-		if (yes_code)
-			if (href_list["time"])
-				var/time = text2num(href_list["time"])
-				timeleft += time
-				timeleft = min(max(round(timeleft), 60), 600)
-			if (href_list["timer"])
-				set_active()
-			if (href_list["safety"])
+	world << action
```
Comment (RemieRichards) [inline code/game/gamemodes/nuclear/nuclearbomb.dm]: ummm

File: code/game/gamemodes/nuclear/nuclearbomb.dm [hunk 2]
```diff
@@ -209,102 +225,129 @@ var/bomb_set
...
+	world << json_encode(params)
```
Comment (RemieRichards) [inline code/game/gamemodes/nuclear/nuclearbomb.dm]: ummmmmmm

**Comment by MrStonedOne** [comment]
>Removes the fact that self destruct nukes are for some reason really high on the layer list?
Thats so when you walk behind it, it layers over you.

File: code/game/gamemodes/nuclear/nuclearbomb.dm [hunk 3]
```diff
@@ -328,36 +373,47 @@ var/bomb_set
...
+		bomb_set = FALSE
+		detonation_timer = null
 		set_security_level(previous_level)
+		countdown.stop()
 	update_icon()
 
+/obj/machinery/nuclearbomb/proc/get_time_left()
+	if(timing)
+		. = max(0, detonation_timer - world.time)
+		. /= 10
```
Comment (RemieRichards) [inline code/game/gamemodes/nuclear/nuclearbomb.dm]: why do this in a separate step?

**Comment by xxalpha** [comment]
The nuke isn't a syndicate device, why does it have the syndicate theme?

**Comment by KorPhaeron** [comment]
Good point!

Please respect our deep lore Coiax

**Comment by WJohn** [comment]
Maybe the syndicate nuke should be syndie themed, while the nanotrasen one is NT themed.

**Comment by MrStonedOne** [comment]
no, the lore is that the syndicate nuke is stolen from NT, hence why it need's NT's disk

**Comment by lzimann** [comment]
Let's please stop adding tgui until #19131 is fixed

**Comment by nullquery** [comment]
#19131 is my bad, I'm releasing a proper fix for it.

**Comment by kilkun** [comment]
The lore is so old it's probably been in the game since before stunsticks replaced police batons. It could do with a rewrite since it doesn't make much sense anyways.

On a more relevant note:

NO. It's one thing to add tgui to shit that doesn't matter like the uplink. Only once in a blue moon do you actually need to open your uplink in a combat situation so it doesn't matter how long it takes to actually open/if it freezes byond. But setting up the nuke is often done in combat. You are practically handing the ops to whatever bloodthirsty greytider in a cardborg suit happens to be on the prowl. Essentially, it's a tangent nerf.

:-1:

also as a general complaint it's butt ugly and all that empty space makes me want to vomit.

**Comment by coiax** [comment]
One of the benefits of TGUI is the seperation of form and function. If someone who understands HTML layouts and shit a little better wants to rearrange the template, they can do so without touching any of the code.

**Comment by kilkun** [comment]
>seperation of form and function

i couldn't care less about doing this even though its staunchly against my principles, but tgui makes the game noticeably suffer. i can deal with the fact that its ugly as sin, i can tolerate it being a slow piece of shit, so long as it doesn't get in the way of when i need to actually do shit. the moment something starts to make the game suffer is the moment you should be weighing just how much it's actually worth in the long run.

adjusting uis is an obscure ease-of-programming thing that very rarely ever gets done. basically giving the crew a free stun when a nuke op opens the bomb menu up is absolutely 100% unacceptable when the reasoning boils down to a slightly abstracted "i dont wanna".

**Comment by coiax** [comment]
~~maybe the op shouldn't be arming the bomb under fire?~~

**Comment by kilkun** [comment]
sometimes thats the only option

**Comment by coiax** [comment]
> i couldn't care less about doing this even though its staunchly against my principles

https://en.wikipedia.org/wiki/Model%E2%80%93view%E2%80%93controller

Better not click on that link fam

**Comment by kilkun** [comment]
I did and now my whole world is upside down.

I'm going to go start a cult devoted to the blood geometer irl complete with self-mutilation in order to avoid processing this turn of events. Goodbye forever.

**Comment by Shadowlight213** [comment]
The annoying thing about tgui, is that it's far harder to understand just by looking at it than html.

While yes, it may be easier for you and others who already know and understand tgui to modify the templates, for others it is much harder, and the tgui template code is absolute gibberish.

**Comment by coiax** [comment]
I don't really have a defence other than, "raw HTML in code is ugly", and "this is just better long term"

**Comment by RemieRichards** [comment]
@Shadowlight213 "absolute gibberish"
{{moustache}} has VERY few rules, and all tgui templates do is combine HTML+moustache for logic, there's also custom tags for tgui such as ```<ui-display>``` and ```<ui-button>``` which just make functionality easier.

**Pull request merged**